tools/halide_image_io.h

```diff
@@ -1048,6 +1048,369 @@ bool save_ppm(ImageType &im, const std::string &filename) {
+        break;
+    case halide_type_t(halide_type_float, 64).as_u32():
+        im->template as<double>().for_each_value([&](double &v) { success &= f.read_bytes(&v, sizeof(v)); });
+        break;
+    case halide_type_t(halide_type_int, 8).as_u32():
+        im->template as<int8_t>().for_each_value([&](int8_t &v) { success &= f.read_bytes(&v, sizeof(v)); });
+        break;
+    case halide_type_t(halide_type_int, 16).as_u32():
+        im->template as<int16_t>().for_each_value([&](int16_t &v) { success &= f.read_bytes(&v, sizeof(v)); });
+        break;
+    case halide_type_t(halide_type_int, 32).as_u32():
+        im->template as<int32_t>().for_each_value([&](int32_t &v) { success &= f.read_bytes(&v, sizeof(v)); });
+        break;
+    case halide_type_t(halide_type_int, 64).as_u32():
+        im->template as<int64_t>().for_each_value([&](int64_t &v) { success &= f.read_bytes(&v, sizeof(v)); });
+        break;
+    case halide_type_t(halide_type_uint, 8).as_u32():
+        im->template as<uint8_t>().for_each_value([&](uint8_t &v) { success &= f.read_bytes(&v, sizeof(v)); });
+        break;
+    case halide_type_t(halide_type_uint, 16).as_u32():
+        im->template as<uint16_t>().for_each_value([&](uint16_t &v) { success &= f.read_bytes(&v, sizeof(v)); });
+        break;
+    case halide_type_t(halide_type_uint, 32).as_u32():
+        im->template as<uint32_t>().for_each_value([&](uint32_t &v) { success &= f.read_bytes(&v, sizeof(v)); });
+        break;
+    case halide_type_t(halide_type_uint, 64).as_u32():
+        im->template as<uint64_t>().for_each_value([&](uint64_t &v) { success &= f.read_bytes(&v, sizeof(v)); });
+        break;
+    default:
+        return check(false, "Unsupported type in load_npy");
+    }
+    if (!check(success, "npy Read Error")) {
+        return false;
+    }
+
+    im->set_host_dirty();
+    return true;
+}
+
+template<typename ImageType, CheckFunc check = CheckReturn>
+bool save_npy(ImageType &im, const std::string &filename) {
+    static_assert(!ImageType::has_static_halide_type, "");
+
+    if (!check(im.copy_to_host() == halide_error_code_success, "copy_to_host() failed.")) {
+        return false;
+    }
+
+    const halide_type_t im_type = im.type();
+    npy_dtype_info_t di = {0, 0, 0};
+    for (const auto &d : npy_dtypes) {
+        if (d.first == im_type) {
+            di = d.second;
+            break;
+        }
+    }
+    if (!check(di.byte_order != 0, "Unsupported type in save_npy")) {
+        return false;
+    }
+
+    std::string shape = "(";
+    for (int d = 0; d < im.dimensions(); ++d) {
+        if (d > 0) {
+            shape += ",";
+        }
+        shape += std::to_string(im.dim(d).extent());
+        if (im.dimensions() == 1) {
+            shape += ",";  // special-case for single-element tuples
+        }
+    }
+    shape += ")";
+
+    // TODO: is it safe to assume that Halide will never write in fortran_order?
```

Comment:
It's not, and for_each_value below doesn't guarantee a planar order. I think you want write_planar_payload

Reply:
Yeah. So, to be clear, if we use `write_planar_payload`, we can definitely assert that fortran_order=false for things we write?

Reply:
Yes, I believe so.